File: Charts/chart.cpp
/*
 This file is part of Marb.

  Marb is free software: you can redistribute it and/or modify
  it under the terms of the Lesser GNU General Public License as published by
  the Free Software Foundation, either version 3 of the License.

  Marb is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
  Lesser GNU General Public License for more details.

  You should have received a copy of the Lesser GNU General Public License
  along with Marb.  If not, see <http://www.gnu.org/licenses/>.

 Marb  Copyright (C) 2013  Dimitry Ernot
*/

#include "chart.h"

#include <QFontMetrics>
#include <qmath.h>
#include <QPen>
#include <QPainter>
#include <QPaintEvent>
#include <QScrollBar>

Chart::Chart( QWidget* parent ) : MarbAbstractItemView( parent ) {
  myMin = 0;
  myMax = 0;
  myMinBound = 0;
  myMaxBound = 0;
  myOrder = 1;
  myNbTicks = 10;
  myTickSize = 0;
  myAlpha = 1;
  myBeta = 1;
  myChartRect = QRect();
  myValuesRect = QRect();
  myTitleRect = QRect();
  myTitle = "";
  myTitleRect = QRect();
  myOrigin = QPoint(0,0);
  myNbDigits = 1;
  myMarginX = 20;
  myMarginY = 20;
}


Chart::~Chart() {
}


void Chart::calculateBounds() {
  myMinBound = myMin;
  myMaxBound = myMax;
  if ( myMaxBound == myMinBound ) {
    ++myMaxBound;
    --myMinBound;
  }
  myOrder = calculateOrder( myMax - myMin );
  myTickSize = (myMax - myMin ) / (myNbTicks - 1);
  if ( myOrder >= 10 ) {
    myNbDigits = 0;
  } else if ( myOrder == 1 ) {
    myNbDigits = 2;
  } else {
    int nbZero = QString::number(myOrder).count( "0" );
    myNbDigits = nbZero + 2;
  }
}


void Chart::calculateLegendRect() {
    QFontMetrics metrics( font() );
    int h = metrics.height() + 5;
    int cols = this->model()->columnCount();
    int nbLines = 1;
    int w = 40;
    int maxWidth = myChartRect.width();
    for( int c = 0; c < cols; ++c ) {
      QString s( this->model()->headerData( c, Qt::Horizontal ).toString() );
      int sWidth = metrics.width( s ) + 40;
      if ( ( w + sWidth ) > maxWidth ) {
        ++nbLines;
        w = sWidth;
      } else {
        w += sWidth;
      }
    }
    myLegendRect = QRect( myMarginX, myMarginY, maxWidth, nbLines * h );
}


qreal Chart::calculateOrder( qreal value ) const {
  qreal order = 1.0;
  qreal v = qAbs( value );
  if ( v >= 1 ) {
    while ( v > 1 ) {
      order *= 10.0;
      v /= 10.0;
    }
    order /= 10.0;
  } else if ( v != 0 ) {
    while ( v < 1 ) {
      order /= 10.0;
      v *= 10.0;
    }
    order *= 10.0;
  }
  return order;
}


ChartStyle Chart::columnStyle( int column ) const {
  if ( myStyle.contains( column ) ) {
      return myStyle[ column ];
  }
  ChartStyle style;
  QColor c1 =Marb::predefinedLightColor( column );
  QColor c2 = Marb::predefinedDarkColor( column );
  style.setPen( QPen( QColor(c2), 2 ) );
  style.setBrush( QBrush(c1) );
  return style;
}

void Chart::defineRects() {
  myChartRect = QRect( QPoint( myMarginX, myMarginY ),
                       this->size() - QSize( myMarginX * 2, myMarginY * 2 ) );

  this->calculateLegendRect();
  myChartRect.setHeight( myChartRect.height() - myLegendRect.height() - 10 );
  myChartRect.translate( 0, myLegendRect.height() + 10 );

  if ( myTitle != "" ) {
    QFont font = this->font();
    font.setItalic( true );
    QFontMetrics metrics( font );
    QRect r( 0, 0, myChartRect.width() - 40, 0 );
    myTitleRect = metrics.boundingRect( r, Qt::AlignHCenter | Qt::AlignTop | Qt::TextWordWrap, myTitle );
    myChartRect.setHeight( myChartRect.height() - myTitleRect.height() - 20 );
  }
}


QModelIndex Chart::indexAt(const QPoint &point) const {
  if ( this->model() == 0 ) {
    return QModelIndex();
  }
  for( int row = 0; row < this->model()->rowCount(); ++row ) {
    for( int col = 0; col < this->model()->columnCount(); ++col ) {
      QModelIndex index = this->model()->index( row, col );
      QPainterPath r = this->itemPath( index );
      if ( r.contains( point ) ) {
        return index;
      }
    }
  }
  return QModelIndex();
}


QPainterPath Chart::itemPath(const QModelIndex &index) const {
  QPainterPath path;
  path.addRect( this->itemRect( index ) );
  return path;
}


/*Paint the legend for the given column. The kind of legend should be defined by each view (linear and radial represent legend differently).
The column legend is represented by a square colored with the pen and brush style and the column name.
*/
void Chart::paintColumnLegend( QPainter& painter, int column, QPoint pos, int metricsH ) {
  QRect r( pos.x() + 20, pos.y() - 10, 20, 20 );
  QPoint posText = pos + QPoint( 45, metricsH/2 );

  ChartStyle style = this->columnStyle( column );

  QString s = this->model()->headerData( column, Qt::Horizontal ).toString();
  painter.drawText( posText, s );
  painter.save();
  painter.setPen( style.pen() );
  painter.setBrush( style.brush() );
  painter.drawRect (r );
  painter.restore();
}


void Chart::paintEvent( QPaintEvent* event ) {
  if ( this->model() == 0 ) {
    return;
  }
    QPainter painter( viewport() );
    painter.setClipRect( event->rect() );
    paintChart( painter );
}


/*Paint the legend in the QRect self._legendRect
The legend corresponds to the text in the Horizontal QHeaderView and the style defined for each column.
*/
void Chart::paintLegend( QPainter& painter) {
    painter.save();
    QFontMetrics metrics( font() );
    int metricsH = metrics.height();
    int h = metricsH + 5;
    int cols = this->model()->columnCount();
    int w = 0;
    int maxWidth = myLegendRect.width();
    QPoint legendPos( myLegendRect.topLeft() );
    QPoint pos = legendPos + QPoint( 50, 0);
    for (int c = 0; c < cols; ++c ) {
      QString s( this->model()->headerData( c, Qt::Horizontal ).toString() );
      int sWidth = metrics.width( s ) + 50;
      QPoint p;
      if ( ( w + sWidth ) > maxWidth ) {
        int y = pos.y();
        p = QPoint( myLegendRect.x() + 10, y + h );
        pos = QPoint( myLegendRect.x() + 50, y + h );
        w = sWidth;
        pos += QPoint( sWidth, 0 );
      } else {
        p = pos + QPoint( -40,  0 );
        w += sWidth;
        pos += QPoint( sWidth, 0 );
      }
      this->paintColumnLegend(painter, c, p + QPoint(0, metricsH), metricsH);
    }
    painter.restore();
}


void Chart::process() {
  if ( this->model() == 0 ) {
    return;
  }
  myMin = 0;
  myMax = 0;
  myMinBottomMargin = this->scanValues() + 10.0;
  this->updateRects();
}


void Chart::resizeEvent(QResizeEvent * ev) {
  QAbstractItemView::resizeEvent( ev );
  this->updateValues();
}


bool Chart::save( QString filename ) {
  QPixmap pix( size() );
  pix.fill( Qt::white );
  QPainter painter( &pix );
  this->paintChart( painter );
  return pix.save( filename );
}


int Chart::scanValues() {
  int rows = this->model()->rowCount();
  int cols = this->model()->columnCount();
  QFontMetrics metrics( font() );
  int textWidth = 0;
  for (int r = 0; r < rows; ++r ) {
    QString s = this->model()->headerData( r, Qt::Vertical ).toString();
    textWidth = qMax( textWidth, metrics.width( s ) + 5 );
    for (int c = 0; c < cols; ++c ) {
      qreal value = this->model()->index( r, c ).data().toReal();
      myMin = qMin( myMin, value );
      myMax = qMax( myMax, value );
    }
  }
  return textWidth;
}


void Chart::setAlphaBeta() {
  myAlpha = -qreal(myValuesRect.height()) / ( myMaxBound - myMinBound );
  myBeta = (myMaxBound * myValuesRect.height() ) / ( myMaxBound - myMinBound ) + myValuesRect.y();
  myOrigin.setY( myBeta );
}


void Chart::setColumnStyle( int column, ChartStyle style ) {
    myStyle[ column ] = style;
}


void Chart::setModel( QAbstractItemModel* model ) {
  QAbstractItemView::setModel( model );
  this->process();
}


void Chart::setScrollBarValues() {
}


void Chart::setTitle( QString title ) {
  myTitle = title;
  this->updateRects();
}


void Chart::updateValues() {
  this->process();
}


qreal Chart::valueToPx( qreal value) const {
  return value * myAlpha + myBeta;
}

File: Marb/Charts/Chart.py
from ..Global import Color, Shape, Type
from PySide.QtGui import QPen, QBrush, \
QColor, QFontMetrics, QPainter, QPixmap, QAbstractItemView
from PySide.QtCore import QRect, QPoint, Qt, QModelIndex, QSize
from ..MarbAbstractItemView import MarbAbstractItemView

from .ChartStyle import ChartStyle

class Axis:
	def __init__( self, nbTicks = 10, minimum = 0, maximum = 0 ):
		self.min = minimum
		self.max = maximum
		self.nbTicks = nbTicks
		self.minBound = 0
		self.maxBound = 0
		self.order = 1
		self.tickSize = 0
		self.alpha = 1
		self.beta = 1
		self.nbDigits = 1
		self.length = 0
		self.axisPos = QPoint(20,10)
		self.origin = QPoint( 0, 0 )


	def calculateBounds(self):
		'''Calculates the minimum bounds and the maximum bounds (i.e., the minimum and the maximum displayed on the chart).
		Calculates the order and the tick size(delta between two ticks on the Y axis) of the charts values.
		If the minimum is equal to the maximum, the minimum bound is equal to minimum - 1 and the maximum bound to maximum + 1
		'''
		self.minBound = self.min
		self.maxBound = self.max
		if abs(self.maxBound - self.minBound) == 0:
			self.maxBound += 1
			self.minBound -= 1
		self.order = self.calculateOrder( self.max - self.min )
		self.tickSize = (self.max - self.min ) / (self.nbTicks - 1)
		if self.order >= 10:
			self.nbDigits = 0
		elif self.order == 1:
			self.nbDigits = 2
		else:
			nbZero = str(self.order).count( '0' )
			self.nbDigits = nbZero + 2

		self.maxBound += self.order
		self.minBound -= self.order

	def calculateOrder( self, value ):
		'''Calculates the order of scale for value (i.e. the power of ten greater than value).
		'''
		order = 1.0
		v = abs( value )
		if v >= 1:
			while v > 1:
				order *= 10.0
				v /= 10.0
			order /= 10.0
		elif v != 0:
			while v < 1:
				order /= 10.0
				v *= 10.0
			order *= 10.0
		return int(order)


	def setAlphaBeta(self, length):
		'''Calculates the coefficients alpha and beta used to convert a value in the model in the corresponding value in pixel in the chart.
		Translates the origin to place the X-Axis.
		'''
		self.length = length
		ln = length * 0.90
		self.alpha = -float( ln ) / float( self.maxBound - self.minBound )
		#self.beta = (self.maxBound * ln ) / ( self.maxBound - self.minBound ) + self.axisPos.y() + length * 0.1
		self.beta = self.maxBound * -self.alpha + self.axisPos.y() + length * 0.1
		self.origin.setY( self.beta )


	def valueToPx(self, value):
		'''Converts the given value in pixel according to the chart scale.
		The pixel value is given by the equation: y = alpha * value + beta 
		'''
		try:
			value = float(value)
		except:
			value = 0
		return value * self.alpha + self.beta


	def paint():
		raise( NotImplementedError, "Must be implemented." )



class Chart(MarbAbstractItemView):
	'''The Chart class provides an abstract base for the chart viewes.
	The Chart class defines the standard interface for every chart views in Marb. It is not supposed to be instantiated directly but should be subclassed.
	'''
	def __init__(self, parent=None):
		'''Constructor. constructs a Chart object with the given 'parent'.
		'''
		super(Chart, self).__init__( parent )
		self.setEditTriggers( QAbstractItemView.NoEditTriggers )
		self._chartRect = QRect()
		self._valuesRect = QRect()
		self._legendRect = QRect()
		self._title = ""
		self._titleRect = QRect()
		self._origin = QPoint(0,0)
		self._style = {}
		self._marginX = 20
		self._marginY = 20
		self.resize( QSize(500, 400) )

		self._yAxis = None


	def calculateLegendRect(self):
		'''Calculates the rectangle needed to display the legend according to the title of each column (given by the QHeaderView).
		'''
		if self.model() == None:
			return None
		metrics = QFontMetrics( self.font() )
		h = metrics.height() + 5
		cols = self.model().columnCount()
		nbLines = 1
		w = 40
		maxWidth = self._chartRect.width()
		for c in range( 0, cols ):
			s = str(self.model().headerData( c, Qt.Horizontal ))
			sWidth = metrics.width( s ) + 40
			if ( w + sWidth ) > maxWidth:
				nbLines += 1
				w = sWidth
			else:
				w += sWidth
		self._legendRect = QRect( self._marginX, self._marginY, maxWidth, nbLines * h )


	def columnStyle(self, column ):
		'''Returns the style for the column.
		If no style has been set for the column, returns the style by default
		'''
		if column in self._style:
			return self._style[ column ]
		else:
			style = ChartStyle()
			c1 = Color.lightColorAt( column )
			c2 = Color.regularColorAt( column )
			style.setPen( QPen( QColor(c2), 2 ) )
			style.setBrush( QBrush(c1) )
			return style


	def defineRects(self):
		''' Define the bounding rectangles used to draw the chart.
		* chartRect: The chart bounds (data + axis)
		* valueRect: Area where the data representation will be drawn. Should be defined inside chartRect
		* legendRect: Area where the legend will be drawn
		* titleRect: Area where the title will be drawn
		'''
		self._chartRect = QRect( QPoint(self._marginX, self._marginY), self.size() - QSize( self._marginX*2, self._marginY*2 ) )
		self.calculateLegendRect()
		self._chartRect.setHeight( self._chartRect.height() - self._legendRect.height() - 10 )
		self._chartRect.translate( 0, self._legendRect.height() + 10 )
		if self._title != "":
			font = self.font()
			font.setItalic( True )
			m = QFontMetrics( font )
			r = QRect( 0, 0, self._chartRect.width() - 40, 0 )
			self._titleRect = m.boundingRect( r, Qt.AlignHCenter | Qt.AlignTop | Qt.TextWordWrap, self._title )
			self._chartRect.setHeight( self._chartRect.height() - self._titleRect.height() - 20 )


	def indexAt(self, point):
		'''Returns the model index of the item at the viewport coordinates point.
		'''
		if self.model() == None:
			return QModelIndex()
		for row in range( self.model().rowCount() ):
			for col in range( self.model().columnCount() ):
				index = self.model().index( row, col )
				r = self.itemPath(index)
				if r.contains( point ):
					return index
		return QModelIndex()


	def paintChart( self, painter ):
		'''Reimplemented.
		Paints the chart itself on the given painter device. Called by paintEvent and save(). 
		'''
		raise( NotImplementedError, "Must be implemented." )


	def _paintColumnLegend(self, painter, c, pos, metricsH):
		'''Paints the legend for the given column. The kind of legend should be defined by each view (linear and radial represent legend differently).
		The column legend is represented by a square colored with the pen and brush style and the column name. 
		'''
		r = QRect( pos.x() + 20, pos.y() - 10, 20, 20 )
		posText = pos + QPoint( 45, metricsH/2 )
		style = self.columnStyle(c)
		s = str(self.model().headerData( c, Qt.Horizontal ))
		painter.drawText( posText, s )
		painter.save()
		painter.setPen( style.pen() )
		painter.setBrush( style.brush() )
		painter.drawRect (r )
		painter.restore()


	def paintEvent( self, event ):
		if self.model() == None:
			return None
		painter = QPainter( self.viewport() )
		painter.setClipRect( event.rect() )
		self.paintChart( painter )


	def _paintLegend(self, painter):
		'''Paint the legend in the QRect self._legendRect
		The legend corresponds to the text in the Horizontal QHeaderView and the style defined for each column. 		
		'''
		painter.save()
		metrics = QFontMetrics( self.font() )
		metricsH = metrics.height()
		h = metricsH + 5
		cols = self.model().columnCount()
		w = 0
		maxWidth = self._legendRect.width()
		legendPos = QPoint(self._legendRect.topLeft()) + QPoint(0, metricsH)
		pos = legendPos + QPoint( 50, 0)
		for c in range( 0, cols ):
			s = str(self.model().headerData( c, Qt.Horizontal ))
			sWidth = metrics.width( s ) + 50
			if ( w + sWidth ) > maxWidth: # Paint on a new line
				y = pos.y()
				p = QPoint( self._legendRect.x() + 10, y + h )
				pos = QPoint( self._legendRect.x() + 50, y + h )
				w = sWidth
				pos += QPoint( sWidth, 0 )
			else:
				p = pos + QPoint( -40, 0 )
				w += sWidth
				pos += QPoint( sWidth, 0 )
			self._paintColumnLegend(painter, c, p, metricsH)
		painter.restore()


	def _paintValues(self, painter, column):
		'''Paints items on the paint device [painter] for the [column]
		'''
		raise( NotImplementedError, "Must be implemented." )

	def process( self ):
		'''Defines the metrics and components to display the chart.
		 Called when model ha changed.
		'''
		if self.model() == None:
			return None
		self._yAxis.min = 0
		self._yAxis.max = 0
		self._updateRects()


	def resizeEvent(self, ev ):
		self.updateValues()
		return MarbAbstractItemView.resizeEvent(self, ev )


	def save(self, filename, size = None ):
		''' Save the chart in the given file filename.
		The file should be an PNG image. 
		'''
		if size == None:
			size = self.size()
		pixmap = QPixmap( size )
		pixmap.fill( Qt.white )
		painter = QPainter( pixmap )
		self.process()
		self.paintChart( painter )
		painter.end()
		return pixmap.save( filename )


	def _scanValues(self):
		'''Scans values in the model to find the minimum and the maximum. Returns the width needed to display the Y scale.
		If the values are greater than zero, the minimum is equal to 0. If the values are less than 0, the maximum is equal to 0.
		If a value is not a number (undefined, a string, etc.), she's considered as equal to 0. 
		'''
		rows = self.model().rowCount()
		cols = self.model().columnCount()
		metrics = QFontMetrics( self.font() )
		textWidth = 0
		# value = self.model().index( 0, 0 ).data()
		# try:
		# 	value = float(value)
		# except:
		# 	value = 0
		_min = 0
		_max = 0
		for r in range( 0, rows ):
			s = str(self.model().headerData( r, Qt.Vertical ))
			textWidth = max( textWidth, metrics.width( s ) + 5 )
			for c in range( 0, cols ):
				value = self.model().index( r, c ).data()
				try:
					value = float(value)
				except:
					value = 0
				_min = float(min( _min, value ))
				_max = float(max( _max, value ))
		self._yAxis.min = _min
		self._yAxis.max = _max
		return textWidth


	def setColumnStyle(self, column, style):
		'''Sets the style for the column.
		'''
		self._style[ column ] = style


	def setTitle(self, title ):
		'''Sets the chart title to title
		Updates the chart to consider the modification
		'''
		if isinstance( title, str):
			self._title = title
			self._updateRects()
			self.update()
		else:
			raise( TypeError, "title must be a string" )


	def _updateRects(self):
		'''Calculates the size and the position of each view part (legend, title,chart)
		Called when the view is resized or model has changed.
		'''
		raise( NotImplementedError, "Must be implemented." )


	def updateValues(self):
		self.process()	


